File: system_simulations/network_model/java/discrete_event/AtomicModel.java
/* Component */
import java.math.BigDecimal;

public interface AtomicModel<Input, Output>{

	public int lambda();

	/* internal state transition function */
	public void deltaInternal();

	/* external state transition function */
	public void deltaExternal(BigDecimal e, int q);

	public void deltaConfluent(int q);

	public BigDecimal timeAdvance();

	public void addInput(Input I);

	public void addOutput(Output O);

	public void log(Event e);

	public Event getLast();

	public String name();
}

File: system_simulations/network_model/cpp/discrete_event/AtomicModel.cpp
#ifndef ATOMICMODEL
#define ATOMICMODEL
#include "Time.cpp"
#include "Scheduler.cpp"
#include <string>
using namespace std;

class AtomicModel {
	public:
		virtual int lambda() = 0;
		virtual void delta_internal() = 0;
		virtual void delta_external(double e, int q) = 0;
		virtual void delta_confluent(int q) = 0;
		virtual double time_advance() = 0;
		virtual string get_name() = 0;
		static Time* current_time;
		Time* last_time;
};

#endif